### ISSUE TITLE
Maintenance Testing: Locales

### ISSUE THREAD

**Issue by jonathanolson** [phetsims/qa#1089]
Testing related to https://github.com/phetsims/joist/issues/963 and https://github.com/phetsims/joist/issues/965.

Locale data for reference: https://github.com/phetsims/babel/blob/main/localeData.json (fixed, so always up-to-date, sorry!)

Related QA dev/main issue: https://github.com/phetsims/qa/issues/1085

This is mainly spot-checking (NOT doing all of the release branches / sims, since that would be impractical).

Behavior to test:

1. Three-letter ?locale query parameter codes (e.g. ?locale=spa) should be equivalent to the equivalent two-character (or extended) code, (e.g. ?locale=es). See the locale data above (anything listed with a locale3 field), examples are en/eng, zho/zh_CN, es/spa.
2. Providing a ?locale value that isn't in the desired format (e.g. xx, xxx, or xx_XX) should pop open the invalid query parameter dialog. Providing any locale IN that format ('x' meaning a-z, 'X' meaning A-Z) should NOT open the dialog, even if it is not a valid locale. Invalid locales (e.g. 'xx' or 'qq') should launch the sim in English.
3. Providing a ?locale value that isn't translated for the sim, AND whose fallbacks are not translated for the sim, will result in the sim loading in English. Fallbacks are in the locale data above (fallbackLocales). Examples are ak_FT: [ ak ], ar_SA: [ ar, ar_MA ], deu: [], pt: [ pt_BR ], es_ES: [ es ].
4. Providing a ?locale value that isn't translated for the sim, BUT a fallback IS translated, will load as the first fallback locale in the list. (can check with visuals of language, or in console with `phet.chipper.locale`).
5. Providing a ?locale value that is translated should load that locale.
6. Locale values in the locale switching part of the settings dialog should be properly sorted (see https://github.com/phetsims/joist/issues/965#issuecomment-2105383072, e.g. Urdu/Persian should not be sorted in weird spots) **In particular, please test this one feature across a range of supported platforms, especially the older ones** (note: dev/main testing took care of the platform-specific part I believe). Additionally, names should match the updated forms from the linked localeData.json (e.g. Zulu will be isiZulu, etc.)
7. String fallback should work, where if a translation does NOT have a translation for a string, but a fallback DOES, the first fallback with a translation will be used for that string. (e.g. in center-and-variability, es_PE itself does not yet have translations for the region-and-culture section (particularly the country strings, e.g. `preferences.tabs.localization.regionAndCulture.unitedStatesOfAmerica`, but the fallback of `es` DOES have translation for this. Additionally, spot-check to make sure there aren't regressions in translation compared to the production sims.
![image](https://github.com/phetsims/qa/assets/480878/0e0fcf01-aec1-48d4-9d16-9e1580b2af02)
8. Dynamic locale changes should work (if it exists for the sim)
9. (phet-io only) phet-io Studio runs, and doesn't complain. `${simName}.general.model.fallbackLocalesProperty`, if it exists (older sims might not have it, if it was before dynamic locale), it should be marked as deprecated in phetioDocumentation, and should NOT be phetioFeatured.
10. (phet-io only) Navigating to a string in studio (while in a particular locale), editing the string should change it in the view. Launching with "Test" should load the simulation, and whenever in that particular (exact!) locale, it should have the changed value.
11. Locale query parameter should be case-insensitive (`ES` or `es_mx` should work), and `-` can be provided as a replacement for _, e.g. `es-mx` should work.
12. (phet-io only). Dynamic locale should be tested according to the instructions in https://github.com/phetsims/phet-io/issues/1881#issuecomment-1287618206.

2024-05-02 balancing-act-1.3 (published, live)

- [x] [balancing-act 1.3.0](https://phet.colorado.edu/sims/html/balancing-act/latest/balancing-act_all.html)

2023-12-04 energy-skate-park-1.3

- [x] [energy-skate-park 1.3.4-rc.1](https://phet-dev.colorado.edu/html/energy-skate-park/1.3.4-rc.1/phet/energy-skate-park_all_phet.html)

2023-04-13 number-compare-1.0

- [x] [number-compare 1.0.7-rc.3](https://phet-dev.colorado.edu/html/number-compare/1.0.7-rc.3/phet/number-compare_all_phet.html)

2022-12-24 ph-scale-1.6

- [x] [ph-scale 1.6.13-rc.1](https://phet-dev.colorado.edu/html/ph-scale/1.6.13-rc.1/phet/ph-scale_all_phet.html)
- [x] [ph-scale 1.6.13-rc.1 phet-io](https://phet-dev.colorado.edu/html/ph-scale/1.6.13-rc.1/phet-io/ph-scale_all_phet-io.html?phetioStandalone)
- [x] [ph-scale 1.6.13-rc.1 phet-io Studio](https://phet-dev.colorado.edu/html/ph-scale/1.6.13-rc.1/phet-io/wrappers/studio)

2022-05-26 function-builder-basics-1.2

- [x] [function-builder-basics 1.2.14-rc.1](https://phet-dev.colorado.edu/html/function-builder-basics/1.2.14-rc.1/phet/function-builder-basics_all_phet.html)

2021-07-22 john-travoltage-1.6

- [x] [john-travoltage 1.6.19-rc.1](https://phet-dev.colorado.edu/html/john-travoltage/1.6.19-rc.1/phet/john-travoltage_all_phet.html)

2020-01-28 gravity-force-lab-2.2

- [x] [gravity-force-lab 2.2.19-rc.1](https://phet-dev.colorado.edu/html/gravity-force-lab/2.2.19-rc.1/phet/gravity-force-lab_all_phet.html)
- [x] [gravity-force-lab 2.2.19-rc.1 phet-io](https://phet-dev.colorado.edu/html/gravity-force-lab/2.2.19-rc.1/phet-io/gravity-force-lab_all_phet-io.html?phetioStandalone)
- [x] [gravity-force-lab 2.2.19-rc.1 phet-io Studio](https://phet-dev.colorado.edu/html/gravity-force-lab/2.2.19-rc.1/phet-io/wrappers/studio)

2019-10-23 vector-addition-equations-1.0

- [x] [vector-addition-equations 1.0.16-rc.1](https://phet-dev.colorado.edu/html/vector-addition-equations/1.0.16-rc.1/phet/vector-addition-equations_all_phet.html)

2019-06-05 fraction-matcher-1.2

- [x] [fraction-matcher 1.2.17-rc.1](https://phet-dev.colorado.edu/html/fraction-matcher/1.2.17-rc.1/phet/fraction-matcher_all_phet.html)

2019-02-01 masses-and-springs-basics-1.0

- [x] [masses-and-springs-basics 1.0.24-rc.1](https://phet-dev.colorado.edu/html/masses-and-springs-basics/1.0.24-rc.1/phet/masses-and-springs-basics_all_phet.html)

2018-07-13 equality-explorer-two-variables-1.0

- [x] [equality-explorer-two-variables 1.0.28-rc.1](https://phet-dev.colorado.edu/html/equality-explorer-two-variables/1.0.28-rc.1/phet/equality-explorer-two-variables_all_phet.html)

2018-01-24 capacitor-lab-basics-1.6

- [x] [capacitor-lab-basics 1.6.35-rc.1](https://phet-dev.colorado.edu/html/capacitor-lab-basics/1.6.35-rc.1/phet/capacitor-lab-basics_all_phet.html)
- [x] [capacitor-lab-basics 1.6.35-rc.1 phet-io](https://phet-dev.colorado.edu/html/capacitor-lab-basics/1.6.35-rc.1/phet-io/capacitor-lab-basics_all_phet-io.html?phetioStandalone)
- [x] [capacitor-lab-basics 1.6.35-rc.1 phet-io Instance Proxies](https://phet-dev.colorado.edu/html/capacitor-lab-basics/1.6.35-rc.1/phet-io/wrappers/instance-proxies/instance-proxies.html?sim=capacitor-lab-basics&relativeSimPath)

2017-11-14 gene-expression-essentials-1.0

- [x] [gene-expression-essentials 1.0.32-rc.2](https://phet-dev.colorado.edu/html/gene-expression-essentials/1.0.32-rc.2/gene-expression-essentials_all.html)

2017-08-28 pendulum-lab-1.0

- [x] [pendulum-lab 1.0.30-rc.2](https://phet-dev.colorado.edu/html/pendulum-lab/1.0.30-rc.2/pendulum-lab_all.html)

2017-01-25 make-a-ten-1.0

- [x] [make-a-ten 1.0.30-rc.2](https://phet-dev.colorado.edu/html/make-a-ten/1.0.30-rc.2/make-a-ten_all.html)

2016-09-14 neuron-1.1

- [x] [neuron 1.1.32-rc.2](https://phet-dev.colorado.edu/html/neuron/1.1.32-rc.2/neuron_all.html)

2016-04-14 energy-skate-park-basics-1.1

- [x] [energy-skate-park-basics 1.1.34-rc.2](https://phet-dev.colorado.edu/html/energy-skate-park-basics/1.1.34-rc.2/energy-skate-park-basics_all.html)

2015-10-01 hookes-law-1.0

- [x] [hookes-law 1.0.38-rc.2](https://phet-dev.colorado.edu/html/hookes-law/1.0.38-rc.2/hookes-law_all.html)

**Comment by KatieWoe** [comment]
Testing step 6:

- [ ] Win 10
- [ ] Mac 10.15
- [ ] Chromebook
- [ ] iOS 15

**Comment by Nancy-Salpepi** [comment]
The parenthesis are incorrect for Arabic locales in pH scale:
<img width="130" alt="Screenshot 2024-05-24 at 4 20 17 PM" src="https://github.com/phetsims/qa/assets/87318828/f58f48a7-12d1-40fa-95bf-4050f8734b2e">

**Comment by KatieWoe** [comment]
The locale parameter doesn't seem to work at all for Hooke's law.
![notatall](https://github.com/phetsims/qa/assets/41024075/25a188fe-548a-4b64-b591-629ebbc8df51)

**Comment by kathy-phet** [comment]
> The parenthesis are incorrect for Arabic locales in pH scale: <img alt="Screenshot 2024-05-24 at 4 20 17 PM" width="130" src="https://private-user-images.githubusercontent.com/87318828/333727228-f58f48a7-12d1-40fa-95bf-4050f8734b2e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1ODI0MDgsIm5iZiI6MTcxNjU4MjEwOCwicGF0aCI6Ii84NzMxODgyOC8zMzM3MjcyMjgtZjU4ZjQ4YTctMTJkMS00MGZhLTk1YmYtNDA1MGY4NzM0YjJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI0VDIwMjE0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQxZTY1NWYwZmYyMjNkNDdkNTEzNDBmOWQxMzlkZGUwMjgxZjM3ODRiZGJlNGMxMGE1NTE0YjlhNjVlOGZlZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DpqFIuqtw_OTaYJZ2Ib9Oo_NPoVLPDGaca69uTYy49w">

@Nancy-Salpepi - This looks platform specific: 
![image](https://github.com/phetsims/qa/assets/5096804/35c37a1b-e558-4da2-8c23-b4008784f63b)

**Comment by KatieWoe** [comment]
On hold while https://github.com/phetsims/qa/issues/1089#issuecomment-2130311183 is investigated

**Comment by jonathanolson** [comment]
I deployed the following RCs (rc.2) with passing "all locales" to the build server. This seems to resolve the issue with the locale query parameter not working.

After talks with MP, I've decided to leave "RC deploys" only using English, due to disk space requirements.

- 2017-11-14 gene-expression-essentials-1.0
- 2017-08-28 pendulum-lab-1.0
- 2017-01-25 make-a-ten-1.0
- 2016-09-14 neuron-1.1
- 2016-04-14 energy-skate-park-basics-1.1
- 2015-10-01 hookes-law-1.0

@KatieWoe I believe I've patched the links in the issue description, presumably this can unblock?

**Comment by jonathanolson** [comment]
> The parenthesis are incorrect for Arabic locales in pH scale: EDIT: I'm on mac + safari (looks correct on mac + chrome)

Looking into it. Looks like that version is inserting non-breaking spaces inside the text.

**Comment by Nancy-Salpepi** [comment]
With mac + safari/chrome
In the list starting with Function Builder (then JT, GFLB)...the fallback locale is working, but in the console when using `phet.chipper.locale` it still lists the original locale. For example, it shows 'de_LI' instead of just 'de' even though there is no translation listed in babel for 'de_LI.' 
The only locales it seems to still work properly for are the ones when the fallback is English (e.g. en_TT will show en in the console).

**Comment by kathy-phet** [comment]
> > The parenthesis are incorrect for Arabic locales in pH scale: EDIT: I'm on mac + safari (looks correct on mac + chrome)
> 
> Tracked it down. I patched this a bit over a year ago in main ([phetsims/joist@286bf32](https://github.com/phetsims/joist/commit/286bf32c1b0def68c3831349dde4838a828cab5c)), haven't heard back about whether it should be maintenance-patched in. @kathy-phet should I apply patches to fix this behavior?

Is it straightforward to add this to the MR? If so, I would say yes. 

For Number Play series, yes, adjust the ordering of the languages to match. 

Thanks!

**Comment by jonathanolson** [comment]
Yes, it should be fairly straightforward to add to this MR.

I'll add both.

**Comment by Nancy-Salpepi** [comment]
@jonathanolson on mac + safari/chrome:
From Vector Addition to the end of the list--I no longer get the invalid query parameter dialog when adding a locale in improper format (behavior 2 in directions). The sim just opens in English.

**Comment by jonathanolson** [comment]
> From Vector Addition to the end of the list--I no longer get the invalid query parameter dialog when adding a locale in improper format (behavior 2 in directions). The sim just opens in English.

It looks like sims at this age and older don't support showing the QueryStringMachine query parameter warnings. So expect the dialog to not open for these (and that is fine/expected).

**Comment by jonathanolson** [comment]
Here are the RCs for the sims affected by the patches for https://github.com/phetsims/qa/issues/1089#issuecomment-2137575237

- number-compare and number-play have their localization settings reordered (no changes to Safari handling, should work)
- the rest have LTR/RTL marks to fix the parentheses in Safari in localization settings.

**beers-law-lab 1.7** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [beers-law-lab 1.7.11-rc.2](https://phet-dev.colorado.edu/html/beers-law-lab/1.7.11-rc.2/phet/beers-law-lab_all_phet.html)
- [ ] [beers-law-lab 1.7.11-rc.2 phet-io](https://phet-dev.colorado.edu/html/beers-law-lab/1.7.11-rc.2/phet-io/beers-law-lab_all_phet-io.html?phetioStandalone)
- [ ] [beers-law-lab 1.7.11-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/beers-law-lab/1.7.11-rc.2/phet-io/wrappers/studio)

**concentration 1.7** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [concentration 1.7.11-rc.2](https://phet-dev.colorado.edu/html/concentration/1.7.11-rc.2/phet/concentration_all_phet.html)
- [ ] [concentration 1.7.11-rc.2 phet-io](https://phet-dev.colorado.edu/html/concentration/1.7.11-rc.2/phet-io/concentration_all_phet-io.html?phetioStandalone)
- [ ] [concentration 1.7.11-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/concentration/1.7.11-rc.2/phet-io/wrappers/studio)

**friction 1.6** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [friction 1.6.12-rc.2](https://phet-dev.colorado.edu/html/friction/1.6.12-rc.2/phet/friction_all_phet.html)
- [ ] [friction 1.6.12-rc.2 phet-io](https://phet-dev.colorado.edu/html/friction/1.6.12-rc.2/phet-io/friction_all_phet-io.html?phetioStandalone)
- [ ] [friction 1.6.12-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/friction/1.6.12-rc.2/phet-io/wrappers/studio)

**gravity-and-orbits 1.6** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [gravity-and-orbits 1.6.19-rc.2](https://phet-dev.colorado.edu/html/gravity-and-orbits/1.6.19-rc.2/phet/gravity-and-orbits_all_phet.html)
- [ ] [gravity-and-orbits 1.6.19-rc.2 phet-io](https://phet-dev.colorado.edu/html/gravity-and-orbits/1.6.19-rc.2/phet-io/gravity-and-orbits_all_phet-io.html?phetioStandalone)
- [ ] [gravity-and-orbits 1.6.19-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/gravity-and-orbits/1.6.19-rc.2/phet-io/wrappers/studio)

**mean-share-and-balance 1.0** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [mean-share-and-balance 1.0.15-rc.2](https://phet-dev.colorado.edu/html/mean-share-and-balance/1.0.15-rc.2/phet/mean-share-and-balance_all_phet.html)

**molecule-shapes 1.6** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [molecule-shapes 1.6.10-rc.2](https://phet-dev.colorado.edu/html/molecule-shapes/1.6.10-rc.2/phet/molecule-shapes_all_phet.html)
- [ ] [molecule-shapes 1.6.10-rc.2 phet-io](https://phet-dev.colorado.edu/html/molecule-shapes/1.6.10-rc.2/phet-io/molecule-shapes_all_phet-io.html?phetioStandalone)
- [ ] [molecule-shapes 1.6.10-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/molecule-shapes/1.6.10-rc.2/phet-io/wrappers/studio)

**molecule-shapes-basics 1.6** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [molecule-shapes-basics 1.6.10-rc.2](https://phet-dev.colorado.edu/html/molecule-shapes-basics/1.6.10-rc.2/phet/molecule-shapes-basics_all_phet.html)
- [ ] [molecule-shapes-basics 1.6.10-rc.2 phet-io](https://phet-dev.colorado.edu/html/molecule-shapes-basics/1.6.10-rc.2/phet-io/molecule-shapes-basics_all_phet-io.html?phetioStandalone)
- [ ] [molecule-shapes-basics 1.6.10-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/molecule-shapes-basics/1.6.10-rc.2/phet-io/wrappers/studio)

**number-compare 1.0** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [number-compare 1.0.7-rc.4](https://phet-dev.colorado.edu/html/number-compare/1.0.7-rc.4/phet/number-compare_all_phet.html)

**number-play 1.1** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [number-play 1.1.7-rc.3](https://phet-dev.colorado.edu/html/number-play/1.1.7-rc.3/phet/number-play_all_phet.html)

**ph-scale 1.6** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [ph-scale 1.6.13-rc.2](https://phet-dev.colorado.edu/html/ph-scale/1.6.13-rc.2/phet/ph-scale_all_phet.html)
- [ ] [ph-scale 1.6.13-rc.2 phet-io](https://phet-dev.colorado.edu/html/ph-scale/1.6.13-rc.2/phet-io/ph-scale_all_phet-io.html?phetioStandalone)
- [ ] [ph-scale 1.6.13-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/ph-scale/1.6.13-rc.2/phet-io/wrappers/studio)

**ph-scale-basics 1.6** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [ph-scale-basics 1.6.12-rc.2](https://phet-dev.colorado.edu/html/ph-scale-basics/1.6.12-rc.2/phet/ph-scale-basics_all_phet.html)
- [ ] [ph-scale-basics 1.6.12-rc.2 phet-io](https://phet-dev.colorado.edu/html/ph-scale-basics/1.6.12-rc.2/phet-io/ph-scale-basics_all_phet-io.html?phetioStandalone)
- [ ] [ph-scale-basics 1.6.12-rc.2 phet-io Studio](https://phet-dev.colorado.edu/html/ph-scale-basics/1.6.12-rc.2/phet-io/wrappers/studio)

**ratio-and-proportion 1.2** (https://github.com/phetsims/joist/issues/963, https://github.com/phetsims/chipper/issues/1379)

- [ ] [ratio-and-proportion 1.2.12-rc.2](https://phet-dev.colorado.edu/html/ratio-and-proportion/1.2.12-rc.2/phet/ratio-and-proportion_all_phet.html)

**Comment by KatieWoe** [comment]
There is a typo in the fallbackLocalesProperty description. "hasguaranteed" should have a space in it. See in pH Scale Studio
>[deprecated, please use localeProperty instead] An ordered list of locales to "fall back" on when a translation is missing for the selected locale, for example: ["es", "de" ]. "en" will always be added to the end of this, because it hasguaranteed full coverage of all translated keys.

**Comment by KatieWoe** [comment]
QA is done

**Comment by jonathanolson** [comment]
> There is a typo in the fallbackLocalesProperty description. "hasguaranteed" should have a space in it. See in pH Scale Studio

Yup, looks like it was fixed in main with this commit (and presumably not patched in release branches): https://github.com/phetsims/joist/commit/5131c263465ab5aebf1808b753447a2fe9fffa50

**Comment by jonathanolson** [comment]
> fallbackLocalesProperty also does not seem to work. Since it is deprecated I'm not sure if this is a bug.

Expected behavior!

**Comment by jonathanolson** [comment]
Going to push out production releases.

**Comment by jonathanolson** [comment]
MR'ed, thank you!
